programs/stake/src/stake_state.rs

```diff
@@ -610,84 +616,56 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
         }
 
         if let StakeState::Uninitialized = split.state()? {
-            // verify enough account lamports
-            if lamports > self.lamports()? {
```

Comment:
hmm, if i read the code correctly, i think this `if` shouldn't be removed as i concerned here a bit (https://github.com/solana-labs/solana/pull/22663/files#r802648796).

Removing this `if` condition will introduce unintended behavior change regarding `StakeState::Uninitialized`.

specifically, newer code will wrongly allow multiple invocations of `StakeInstruction::Split`s on **0-lamport** already `Uninitialized` stake accounts as `self` here in a single tx. That behavior itself is a meaningless and no-op. However, older code will abort at the 2nd `Split` instruction immediately. Thus resulting in tx status outcome change.

That's because the subsequent codepath for `StakeState::Uninitialized` doesn't prevent such an nonsense action and just run as no-op, after removing this (if i read the code).

Maybe, we should restore this check here because it's rather general check like the preceeding `.owner()` and `.data_len()`. it's easy to miss all of following `match` arms are properly guarded as before like when adding newer `StakeState` variant in the future (quite unlikely though).

Could you double-check this? and add test to demonstrate this unintended behavior change?

Reply:
@ryoqun Can you help me with the test procedure? Right now I have basically this:

```
- source account set to `Uninitialized` with zero lamports
- dest1 account set to `Uninitialized` with zero lamports
- split source into dest1 with amount = 0; assert Ok
- dest2 account set to `Uninitialized` with zero lamports
- split source into dest2 with amount = 0; assert Err
```

And that returns Ok for both asserts, for both my PR and against master. So I'm guessing I must be missing something.

<details>
<summary>Here's the actual code</summary>

```rust
    #[test]
    fn test_multiple_splits_from_uninitialized() {
        let source_pubkey = Pubkey::new_unique();
        let source_account = AccountSharedData::new_ref_data_with_space(
            0,
            &StakeState::Uninitialized,
            std::mem::size_of::<StakeState>(),
            &id(),
        )
        .unwrap();
        let source_keyed_account = KeyedAccount::new(&source_pubkey, true, &source_account);

        for i in 0..10 {
            let dest_pubkey = Pubkey::new_unique();
            let dest_account = AccountSharedData::new_ref_data_with_space(
                0,
                &StakeState::Uninitialized,
                std::mem::size_of::<StakeState>(),
                &id(),
            )
            .unwrap();
            let dest_keyed_account = KeyedAccount::new(&dest_pubkey, true, &dest_account);

            let result =
                source_keyed_account.split(0, &dest_keyed_account, &HashSet::from([source_pubkey]));
            assert!(result.is_ok());
        }
    }
```

And the test passes for both master and this PR, with the 10 splits.
</details>

Reply:
My read is that the only way to get success with `lamports > self.lamports()?` is if `self.key == split.key` because otherwise `self.try_account_ref_mut()?.checked_sub_lamports(lamports)?` would fail due to underflow. It looks like the change in this PR allows such a transaction to succeed.

Reply:
gah, as @jstarry said, i lacked the important bit to decipher my previous comment... yeah, i definitely should had mentioned about `self-split`... sorry about confusion.

Reply:
Oh, duh, hah I see now. I forgot that `validate_split_amount()` was moved into the match arms... I've re-added the check _above_ the match.

For the test, I still always get `split()` to return an error if `lamports > self.lamports()`, specifically because of the `.checked_sub_lamports()` returning an Underflow error. And calling `split()` with zero lamports always succeeds (on master)...
I feel like I'm still not understanding something properly though.

Reply:
(hmm, i'll check this shortly after).

Reply:
@brooksprumo my bad. my above comment had bugs (specifically, the 2nd ix and 0-lamport requirement were total false information). maybe i should just write down the code instead of english.. ;)

here's the updated test, which should demonstrate the differing behavior: 
https://github.com/ryoqun/solana/tree/minimum-stake-constant-test-tweaks

also, your latest pushed commits looks okay.

Reply:
@ryoqun Thanks for the code! I should've replied back here last night; @jstarry pointed out my issue, needing to use the same keyed account for both source and destination of the split.

Reply:
@brooksprumo no problem. could you add this new test into pr for future prosperity? i think we just lacked another edge case here, resulting in unnoticed behavior change.

Reply:
@ryoqun Yep on it!

Reply:
@ryoqun Tests are over here: https://github.com/solana-labs/solana/pull/23442

Reply:
I've also run this branch against those tests. So this PR should be safe to review again. I can ensure the other PR is merged first, and this PR gets rebased and re-runs CI as well.

Reply:
Done! Tests have been merged and this PR has been rebased.